File: backend-java/src/integrationTest/java/com/bffgin/backend/external/ExternalHandlerIntegrationTest.java
package com.bffgin.backend.external;

import com.bffgin.backend.DbTestFixture;
import com.bffgin.backend.auth.Dispatcher;
import com.bffgin.backend.auth.HmacVerifier;
import com.bffgin.backend.auth.JwksVerifier;
import com.bffgin.backend.domain.TaskInput;
import com.bffgin.backend.domain.TaskStatus;
import com.bffgin.backend.flags.FeatureFlagPoller;
import com.bffgin.backend.repository.TaskRepository;
import com.bffgin.backend.test_support.MockJwksServer;
import com.bffgin.backend.test_support.TestTokenHelper;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.javalin.Javalin;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.security.KeyPair;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.time.LocalDate;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * 実DB(docker-compose上のMySQL)+実HTTPサーバー+モックJWKSサーバーによる、外部公開API
 * (/external/v1/tasks)の結合テスト。TaskGrpcIntegrationTestと同じ「クラス全体で1つの
 * 実サーバーを起動する」設計。モックJWKSサーバーをKeycloak相当として扱う
 * (backend-c/backend-cppのexternal_handler_integration_testと同じ設計)
 */
class ExternalHandlerIntegrationTest {

    private static final String EXTERNAL_CLIENT_ID = "external-api-client";
    private static final String KEYCLOAK_ISSUER = "https://keycloak.example/realms/training";
    private static final String LOCAL_HMAC_SECRET = "test-secret-at-least-32-bytes-long!!";
    private static final int TEST_PORT = 18112;

    private static MockJwksServer mockJwks;
    private static KeyPair keyPair;
    private static FeatureFlagPoller flagPoller;
    private static Javalin app;
    private static DbTestFixture fixture;
    private static final HttpClient httpClient = HttpClient.newHttpClient();
    private static final ObjectMapper mapper = new ObjectMapper();

    private long userId;
    private String suffix;

    @BeforeAll
    static void startServer() throws Exception {
        fixture = new DbTestFixture();
        TaskRepository repository = new TaskRepository(fixture.dataSource);

        keyPair = TestTokenHelper.generateRsaKeyPair();
        mockJwks = new MockJwksServer();
        mockJwks.addKey("kid-1", (RSAPublicKey) keyPair.getPublic());

        Dispatcher dispatcher = new Dispatcher()
                .register(Dispatcher.LOCAL_HMAC_ISSUER,
                        new HmacVerifier(LOCAL_HMAC_SECRET, Dispatcher.LOCAL_HMAC_ISSUER, "backend"))
                .register(KEYCLOAK_ISSUER, new JwksVerifier(mockJwks.jwksUrl(), KEYCLOAK_ISSUER, "backend"));

        flagPoller = new FeatureFlagPoller(fixture.dataSource);
        flagPoller.pollOnce();

        ExternalHandler handler = new ExternalHandler(repository, dispatcher, flagPoller, EXTERNAL_CLIENT_ID);
        app = Javalin.create();
        app.get("/external/v1/tasks", handler.list);
        app.exception(com.bffgin.backend.domain.TaskError.class,
                (e, ctx) -> com.bffgin.backend.rest.RestErrorMapper.write(ctx, e));
        app.start(TEST_PORT);
    }

    @AfterAll
    static void stopServer() {
        app.stop();
        mockJwks.close();
        fixture.close();
    }

    @BeforeEach
    void setUp() throws Exception {
        suffix = System.nanoTime() + "-" + Math.abs(new java.util.Random().nextInt());
        userId = fixture.createUser(suffix);
    }

    @AfterEach
    void tearDown() {
        fixture.cleanupUser(userId);
    }

    private String keycloakToken() {
        return TestTokenHelper.makeRsaToken((RSAPrivateKey) keyPair.getPrivate(), "kid-1", KEYCLOAK_ISSUER, "backend",
                "some-keycloak-sub-" + suffix, 3600, "RS256", EXTERNAL_CLIENT_ID);
    }

    private HttpResponse<String> get(String path, String bearer) throws Exception {
        HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create("http://127.0.0.1:" + TEST_PORT + path)).GET();
        if (bearer != null) {
            builder.header("Authorization", "Bearer " + bearer);
        }
        return httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
    }

    @Test
    void rejectsMissingAuthorization() throws Exception {
        var resp = get("/external/v1/tasks?user_id=" + userId, null);
        assertEquals(401, resp.statusCode());
    }

    @Test
    void rejectsWrongAzp() throws Exception {
        String token = TestTokenHelper.makeRsaToken((RSAPrivateKey) keyPair.getPrivate(), "kid-1", KEYCLOAK_ISSUER,
                "backend", "sub", 3600, "RS256", "some-other-client");
        var resp = get("/external/v1/tasks?user_id=" + userId, token);
        assertEquals(401, resp.statusCode());
    }

    @Test
    void rejectsLocalHmacIssuerEvenWithCorrectAzp() throws Exception {
        String token = TestTokenHelper.makeHmacToken(LOCAL_HMAC_SECRET, Dispatcher.LOCAL_HMAC_ISSUER, "backend",
                String.valueOf(userId), 3600, "HS256", EXTERNAL_CLIENT_ID);
        var resp = get("/external/v1/tasks?user_id=" + userId, token);
        assertEquals(401, resp.statusCode());
    }

    @Test
    void rejectsMissingUserId() throws Exception {
        var resp = get("/external/v1/tasks", keycloakToken());
        assertEquals(400, resp.statusCode());
        JsonNode body = mapper.readTree(resp.body());
        assertEquals("user_id_required", body.get("error").asText());
    }

    @Test
    void rejectsInvalidUserId() throws Exception {
        var resp = get("/external/v1/tasks?user_id=not-a-number", keycloakToken());
        assertEquals(400, resp.statusCode());
        JsonNode body = mapper.readTree(resp.body());
        assertEquals("invalid_user_id", body.get("error").asText());
    }

    @Test
    void offsetPaginationAcrossPages() throws Exception {
        TaskRepository repository = new TaskRepository(fixture.dataSource);
        for (int i = 0; i < 3; i++) {
            repository.create(userId,
                    new TaskInput("task-" + i, "desc", "waiting", LocalDate.of(2099, 1, 1), List.of()),
                    TaskStatus.WAITING);
        }

        var page1 = get("/external/v1/tasks?user_id=" + userId + "&page=1&page_size=2", keycloakToken());
        assertEquals(200, page1.statusCode());
        JsonNode body1 = mapper.readTree(page1.body());
        assertEquals(2, body1.get("tasks").size());
        assertEquals(1, body1.get("page").asInt());
        assertEquals(2, body1.get("page_size").asInt());
        assertEquals(3, body1.get("total").asInt());

        var page2 = get("/external/v1/tasks?user_id=" + userId + "&page=2&page_size=2", keycloakToken());
        JsonNode body2 = mapper.readTree(page2.body());
        assertEquals(1, body2.get("tasks").size());
    }

    @Test
    void cursorPaginationChainsToNull() throws Exception {
        try {
            setFlag(true, "on");

            TaskRepository repository = new TaskRepository(fixture.dataSource);
            long id1 = repository.create(userId,
                    new TaskInput("cursor-task-1", "desc", "waiting", LocalDate.of(2099, 1, 1), List.of()),
                    TaskStatus.WAITING);
            long id2 = repository.create(userId,
                    new TaskInput("cursor-task-2", "desc", "waiting", LocalDate.of(2099, 1, 1), List.of()),
                    TaskStatus.WAITING);
            flagPoller.pollOnce();

            var page1 = get("/external/v1/tasks?user_id=" + userId + "&limit=1", keycloakToken());
            assertEquals(200, page1.statusCode());
            JsonNode body1 = mapper.readTree(page1.body());
            assertEquals(1, body1.get("tasks").size());
            assertEquals(String.valueOf(id1), body1.get("next_cursor").asText());

            var page2 = get("/external/v1/tasks?user_id=" + userId + "&cursor=" + id1 + "&limit=1", keycloakToken());
            JsonNode body2 = mapper.readTree(page2.body());
            assertEquals(1, body2.get("tasks").size());
            assertEquals(String.valueOf(id2), body2.get("next_cursor").asText());

            var page3 = get("/external/v1/tasks?user_id=" + userId + "&cursor=" + id2 + "&limit=1", keycloakToken());
            JsonNode body3 = mapper.readTree(page3.body());
            assertEquals(0, body3.get("tasks").size());
            assertTrue(body3.get("next_cursor").isNull());
        } finally {
            setFlag(false, "off");
            flagPoller.pollOnce();
        }
    }

    /**
     * backend.external-tasks-pagination-v2は全言語のテストスイートで共有されるフラグ行のため、
     * 変更前の状態を必ず復元する(finally節、他言語の結合テストと同じ方針)
     */
    private void setFlag(boolean enabled, String defaultVariation) throws Exception {
        try (Connection conn = fixture.dataSource.getConnection();
                PreparedStatement ps = conn.prepareStatement(
                        "UPDATE feature_flags SET enabled = ?, default_variation = ? WHERE flag_key = 'backend.external-tasks-pagination-v2'")) {
            ps.setBoolean(1, enabled);
            ps.setString(2, defaultVariation);
            int updated = ps.executeUpdate();
            assertTrue(updated > 0, "backend.external-tasks-pagination-v2 flag row must already exist");
        }
    }
}

File: backend-kotlin/src/integrationTest/kotlin/com/bffgin/backend/rest/ExternalHandlerIntegrationTest.kt
package com.bffgin.backend.rest

import com.bffgin.backend.DbTestFixture
import com.bffgin.backend.auth.Dispatcher
import com.bffgin.backend.domain.TaskInput
import com.bffgin.backend.domain.TaskStatus
import com.bffgin.backend.flags.FeatureFlagPoller
import com.bffgin.backend.repository.TaskRepository
import com.bffgin.backend.test_support.MockJwksServer
import com.bffgin.backend.test_support.TestTokenHelper
import com.fasterxml.jackson.databind.ObjectMapper
import io.ktor.server.cio.CIO
import io.ktor.server.engine.ApplicationEngine
import io.ktor.server.engine.embeddedServer
import io.ktor.server.routing.routing
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.time.LocalDate

/**
 * 実DB・モックJWKSサーバー(Keycloak相当として扱う)・外部公開API自身のKtorリスナー
 * (テスト用ポート)の3つを1プロセス内に立て、実際にHTTP GETを送ってエンドツーエンドに
 * 検証する(backend-c/backend-cpp/backend-javaのexternal_handler_integration_test相当)。
 * backend.external-tasks-pagination-v2は退避→書き換え→テスト後に復元する
 * (全言語で共有する1つのFeature Flagのため、他のテスト・実機動作に影響を残さない)
 */
class ExternalHandlerIntegrationTest {

    private val externalApiClientId = "external-api-client"
    private val keycloakIssuer = "https://keycloak.example/realms/training"
    private val flagKey = "backend.external-tasks-pagination-v2"

    private lateinit var fixture: DbTestFixture
    private lateinit var repository: TaskRepository
    private lateinit var mockJwks: MockJwksServer
    private lateinit var flagPoller: FeatureFlagPoller
    private lateinit var scope: CoroutineScope
    private lateinit var server: ApplicationEngine
    private var port: Int = 0
    private lateinit var keyPair: java.security.KeyPair
    private var userId: Long = 0
    private var originalFlagEnabled: Boolean = false
    private var originalFlagDefault: String = "off"

    private val http = HttpClient.newHttpClient()
    private val mapper = ObjectMapper()

    @BeforeEach
    fun setUp() {
        fixture = DbTestFixture()
        repository = TaskRepository(fixture.dataSource)
        userId = fixture.createUser("${System.nanoTime()}")

        mockJwks = MockJwksServer()
        keyPair = TestTokenHelper.generateRsaKeyPair()
        mockJwks.addKey("kid-1", keyPair.public as RSAPublicKey)

        val dispatcher = Dispatcher().register(
            keycloakIssuer,
            com.bffgin.backend.auth.JwksVerifier(mockJwks.jwksUrl(), keycloakIssuer, "backend"),
        )

        scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        flagPoller = FeatureFlagPoller(fixture.dataSource)
        flagPoller.start(scope)

        // 実行中の値を退避(テスト後に復元する)
        fixture.dataSource.connection.use { conn ->
            conn.prepareStatement("SELECT enabled, default_variation FROM feature_flags WHERE flag_key = ?").use { ps ->
                ps.setString(1, flagKey)
                ps.executeQuery().use { rs ->
                    if (rs.next()) {
                        originalFlagEnabled = rs.getBoolean("enabled")
                        originalFlagDefault = rs.getString("default_variation")
                    }
                }
            }
        }

        port = findFreePort()
        server = embeddedServer(CIO, port = port) {
            routing {
                externalRoutes(repository, dispatcher, flagPoller, externalApiClientId)
            }
        }
        server.start(wait = false)
        // サーバー起動を待つ(即座にリクエストすると接続拒否になりうる)
        Thread.sleep(300)
        runBlocking { flagPoller.pollOnce() }
    }

    @AfterEach
    fun tearDown() {
        try {
            restoreFlag()
        } finally {
            server.stop(0, 0)
            scope.cancel()
            mockJwks.close()
            fixture.cleanupUser(userId)
            fixture.close()
        }
    }

    private fun findFreePort(): Int {
        java.net.ServerSocket(0).use { return it.localPort }
    }

    private fun setFlag(enabled: Boolean, defaultVariation: String) {
        fixture.dataSource.connection.use { conn ->
            conn.prepareStatement("UPDATE feature_flags SET enabled = ?, default_variation = ? WHERE flag_key = ?").use { ps ->
                ps.setBoolean(1, enabled)
                ps.setString(2, defaultVariation)
                ps.setString(3, flagKey)
                ps.executeUpdate()
            }
        }
    }

    private fun restoreFlag() {
        setFlag(originalFlagEnabled, originalFlagDefault)
    }

    private fun validToken(): String = TestTokenHelper.makeRsaToken(
        keyPair.private as RSAPrivateKey, "kid-1", keycloakIssuer, "backend", "1", 3600,
        azp = externalApiClientId,
    )

    private fun get(path: String, token: String? = null): HttpResponse<String> {
        val builder = HttpRequest.newBuilder(URI.create("http://127.0.0.1:$port$path")).GET()
        if (token != null) {
            builder.header("Authorization", "Bearer $token")
        }
        return http.send(builder.build(), HttpResponse.BodyHandlers.ofString())
    }

    @Test
    fun rejectsMissingAuthorization() {
        val response = get("/external/v1/tasks?user_id=$userId")
        assertEquals(401, response.statusCode())
        assertEquals("unauthenticated", mapper.readTree(response.body()).get("error").asText())
    }

    @Test
    fun rejectsWrongAzp() {
        val token = TestTokenHelper.makeRsaToken(
            keyPair.private as RSAPrivateKey, "kid-1", keycloakIssuer, "backend", "1", 3600,
            azp = "wrong-client",
        )
        val response = get("/external/v1/tasks?user_id=$userId", token)
        assertEquals(401, response.statusCode())
    }

    @Test
    fun rejectsMissingUserId() {
        val response = get("/external/v1/tasks", validToken())
        assertEquals(400, response.statusCode())
        assertEquals("user_id_required", mapper.readTree(response.body()).get("error").asText())
    }

    @Test
    fun offsetPaginationAcrossPages() = runBlocking {
        val labelId = fixture.createLabel("${System.nanoTime()}")
        try {
            repository.create(userId, TaskInput("t1", null, "waiting", LocalDate.of(2099, 1, 1), listOf(labelId)), TaskStatus.WAITING)
            repository.create(userId, TaskInput("t2", null, "waiting", LocalDate.of(2099, 1, 1), listOf(labelId)), TaskStatus.WAITING)
            repository.create(userId, TaskInput("t3", null, "waiting", LocalDate.of(2099, 1, 1), listOf(labelId)), TaskStatus.WAITING)

            setFlag(false, "off")

            val page1 = get("/external/v1/tasks?user_id=$userId&page=1&page_size=2", validToken())
            assertEquals(200, page1.statusCode())
            val body1 = mapper.readTree(page1.body())
            assertEquals(2, body1.get("tasks").size())
            assertEquals(3, body1.get("total").asInt())
            assertEquals(1, body1.get("page").asInt())
            assertEquals(2, body1.get("page_size").asInt())

            val page2 = get("/external/v1/tasks?user_id=$userId&page=2&page_size=2", validToken())
            val body2 = mapper.readTree(page2.body())
            assertEquals(1, body2.get("tasks").size())
        } finally {
            fixture.cleanupLabel(labelId)
        }
    }

    @Test
    fun cursorPaginationChainsToNull() = runBlocking {
        val labelId = fixture.createLabel("${System.nanoTime()}")
        try {
            val id1 = repository.create(userId, TaskInput("c1", null, "waiting", LocalDate.of(2099, 1, 1), listOf(labelId)), TaskStatus.WAITING)
            val id2 = repository.create(userId, TaskInput("c2", null, "waiting", LocalDate.of(2099, 1, 1), listOf(labelId)), TaskStatus.WAITING)

            setFlag(true, "on")
            runBlocking { flagPoller.pollOnce() }

            val page1 = get("/external/v1/tasks?user_id=$userId&limit=1", validToken())
            assertEquals(200, page1.statusCode())
            val body1 = mapper.readTree(page1.body())
            assertEquals(1, body1.get("tasks").size())
            assertEquals(id1.toString(), body1.get("next_cursor").asText())

            val page2 = get("/external/v1/tasks?user_id=$userId&cursor=$id1&limit=1", validToken())
            val body2 = mapper.readTree(page2.body())
            assertEquals(1, body2.get("tasks").size())
            assertEquals(id2.toString(), body2.get("next_cursor").asText())

            val page3 = get("/external/v1/tasks?user_id=$userId&cursor=$id2&limit=1", validToken())
            val body3 = mapper.readTree(page3.body())
            assertEquals(0, body3.get("tasks").size())
            assertTrue(body3.get("next_cursor").isNull)
        } finally {
            fixture.cleanupLabel(labelId)
        }
    }
}
